src/shogun/optimization/FirstOrderSAGCostFunctionInterface.cpp

```diff
@@ -0,0 +1,178 @@
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ *
+ */
+
+ #include <shogun/optimization/FirstOrderSAGCostFunctionInterface.h>
+ using namespace shogun;
+
+ FirstOrderSAGCostFunctionInterface::FirstOrderSAGCostFunctionInterface(
+   Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic>* X,
+   Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y,
+   Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* trainable_parameters,
+   std::vector< std::function < stan::math::var(int32_t) > >* cost_for_ith_point,
+   std::function < stan::math::var(std::vector<stan::math::var>*) >* total_cost
+ )
+ {
+   m_X                    = X;
+   m_y                    = y;
+   m_trainable_parameters = trainable_parameters;
+   m_cost_for_ith_point   = cost_for_ith_point;
+   m_total_cost           = total_cost;
+ }
+
+ void FirstOrderSAGCostFunctionInterface::set_training_data(
+   Eigen::Matrix<float64_t, Eigen::Dynamic, Eigen::Dynamic> * X_new,
+   Eigen::Matrix<float64_t, 1, Eigen::Dynamic>* y_new
+ )
+ {
+  REQUIRE(X_new!=NULL, "X must be non NULL");
+  REQUIRE(y_new!=NULL, "Y must be non NULL");
+  if (this->m_X != X_new)
+  {
+   this->m_X = X_new;
+  }
+  if (this->m_y != y_new)
+  {
+   this->m_y = y_new;
+  }
+ }
+
+ void FirstOrderSAGCostFunctionInterface::set_trainable_parameters(
+   Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>* new_params
+ )
+ {
+   REQUIRE(new_params, "The trainable parameters must be non NULL");
+   if (this->m_trainable_parameters != new_params)
+   {
+     this->m_trainable_parameters = new_params;
+   }
+ }
+
+ void FirstOrderSAGCostFunctionInterface::set_ith_cost_function(
+   std::vector< std::function < stan::math::var(int32_t) > >* new_cost_f
+ )
+ {
+   REQUIRE(new_cost_f, "The cost function must be a non NULL vector of stan variables");
+   if (this->m_cost_for_ith_point != new_cost_f)
+   {
+     this->m_cost_for_ith_point = new_cost_f;
+   }
+ }
+
+ void FirstOrderSAGCostFunctionInterface::set_cost_function(
+   std::function < stan::math::var(std::vector<stan::math::var>*) >* total_cost
+ )
+ {
+   REQUIRE(total_cost, "The total cost function must be a non NULL function returning a stan variable");
+   if (this->m_total_cost != total_cost)
+   {
+     this->m_total_cost = total_cost;
+   }
+ }
+
+ FirstOrderSAGCostFunctionInterface::~FirstOrderSAGCostFunctionInterface()
+ {
+ }
+
+ void FirstOrderSAGCostFunctionInterface::begin_sample()
+ {
+   m_index_of_sample = 0;
+ }
+
+ bool FirstOrderSAGCostFunctionInterface::next_sample()
+ {
+  int32_t num_of_samples = m_X->cols();
+  if (m_index_of_sample>=num_of_samples)return false;
+  ++m_index_of_sample;
+  return true;
+ }
+
+
+SGVector<float64_t> FirstOrderSAGCostFunctionInterface::get_gradient()
+{
+  int32_t num_of_variables = m_trainable_parameters->rows();
+  REQUIRE(num_of_variables > 0, "Number of training parameters must be greater than 0");
```

Comment:
Pls always print the provided number
"Number of training parameters (%d) must be positive.\n"
https://github.com/shogun-toolbox/shogun/wiki/Assertions